### PR TITLE
feat(integration): Singular- unixtimestamp format fixes and empty url allowed for openuri

### PR DESCRIPTION
**Pull request by yashasvibajpai**
## Description of the change

> `utime`, `install_time` and `update_time` parameters converted into unixtimestamp format (seconds epoch).
> Empty string is now default value for openuri

## Type of change
- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: v0/destinations/singular/util.js [hunk 1]
```diff
@@ -138,6 +139,26 @@ const platformWisePayloadGenerator = (message, isSessionEvent) => {
...
+    const it =
+      message.properties?.install_time ||
+      message.originalTimestamp ||
+      message.timestamp;
+    const ut =
+      message.properties?.update_time ||
+      message.originalTimestamp ||
+      message.timestamp;
+    payload.install_time = toUnixTimestamp(it);
+    payload.update_time = toUnixTimestamp(ut);
```
Comment (utsabc) [inline v0/destinations/singular/util.js:154]: Why cant we do this using mapping json. We can have this `toUnixTimestamp` as metadata function in mapping json itself right?

File: v0/destinations/singular/util.js [hunk 2]
```diff
@@ -161,6 +182,11 @@ const platformWisePayloadGenerator = (message, isSessionEvent) => {
   } else {
     payload.c = "carrier";
   }
+
+  // Singular accepts utime values in Unix timestamp, hence we are converting it
+  const ts = message.originalTimestamp || message.timestamp;
+  payload.utime = toUnixTimestamp(ts);
```
Comment (utsabc) [inline v0/destinations/singular/util.js:188]: Same here if we can use mapping JSON

File: v0/destinations/singular/util.js [hunk 3]
```diff
@@ -138,6 +139,26 @@ const platformWisePayloadGenerator = (message, isSessionEvent) => {
...
+      message.timestamp;
+    payload.install_time = toUnixTimestamp(it);
+    payload.update_time = toUnixTimestamp(ut);
+
+    // set openuri in payload as "" (empty string) if url is not passed in properties
+    if (!message.properties.url) {
+      payload.openuri = "";
+    } else {
+      payload.openuri = message.properties.url;
+    }
```
Comment (utsabc) [inline v0/destinations/singular/util.js:161]:
```suggestion
    payload.openuri = message.properties.url || "";
```

Comment (utsabc) [reply v0/destinations/singular/util.js:161]: Similarly can we check other fields as well?
Screenshot 2022-10-12 at 10 44 51 AM

File: v0/destinations/singular/util.js (hunk 1 above)
Comment (yashasvibajpai) [reply v0/destinations/singular/util.js:154]: reinstated mapping and used metadata

File: v0/destinations/singular/util.js (hunk 2 above)
Comment (yashasvibajpai) [reply v0/destinations/singular/util.js:188]: done

File: v0/destinations/singular/util.js (hunk 1 above)
Comment (utsabc) [reply v0/destinations/singular/util.js:154]: Not seeing can you push your changes

File: v0/destinations/singular/util.js (hunk 3 above)
Comment (yashasvibajpai) [reply v0/destinations/singular/util.js]: We are removing all null and undefined values from the payload in the end.

File: v0/destinations/singular/util.js [hunk 4]
```diff
@@ -161,6 +164,11 @@ const platformWisePayloadGenerator = (message, isSessionEvent) => {
   } else {
     payload.c = "carrier";
   }
+
+  // Singular accepts utime values in Unix timestamp, hence we are converting it
+  const ts = message.originalTimestamp || message.timestamp;
+  payload.utime = toUnixTimestamp(ts);
```
Comment (utsabc) [inline v0/destinations/singular/util.js:170]: We don't need this right?

File: v0/destinations/singular/util.js [hunk 5]
```diff
@@ -138,6 +139,8 @@ const platformWisePayloadGenerator = (message, isSessionEvent) => {
     } else {
       payload.dnt = 1;
     }
+    // by default, the value of openuri should be "", i.e empty string if nothing is passed
+    payload.openuri = message.properties.url || "";
```
Comment (utsabc) [inline v0/destinations/singular/util.js:143]: This is same as
```
{
    "destKey": "openuri",
    "sourceKeys": "properties.url",
    "required": true,
    "metadata": {
      "defaultValue": ""
    }
  }
  ```
  Shall we use mapping json way only?

Comment (yashasvibajpai) [reply v0/destinations/singular/util.js:143]: Will "defaultValue" work with "required" mappings? I didn't see it working, that's why used the other way.

File: v0/destinations/singular/util.js (hunk 4 above)
Comment (yashasvibajpai) [reply v0/destinations/singular/util.js:170]: No, removing it

File: v0/destinations/singular/util.js (hunk 5 above)
Comment (utsabc) [reply v0/destinations/singular/util.js:143]: We can remove "required" then

Comment (yashasvibajpai) [reply v0/destinations/singular/util.js:143]: It is not working.

Comment (utsabc) [reply v0/destinations/singular/util.js:143]: ok then its fine with current implementation

**Comment by utsabc** [comment]
@yashasvibajpai can you get 1 more approval from the team

**Review by ujjwal-ab — APPROVED** [review]

**Pull request merged**